### ISSUE TITLE
Feature Request: Advanced Indexing

### ISSUE THREAD

**Issue by zuoxingdong** [pytorch/pytorch#1080]
For the following case, `torch.gather` does not work in one line

```
a = np.zeros([5, 4, 3])
a[range(5), :, [0, 1, 1, 2, 0]].shape
```

Hope this functionality could be added. e.g. `gather_n`

**Comment by fmassa** [comment]
I have a prototype in python that implements advanced indexing according to numpy rules.
@apaszke you can assign this to me

**Comment by zuoxingdong** [comment]
@fmassa Can it work with `backward()` and on cuda ?

**Comment by fmassa** [comment]
On CUDA, yes.
With some work it can also for backward.

**Comment by zuoxingdong** [comment]
@fmassa That's great ! Could you submit a PR ?

**Comment by soumith** [comment]
@zuoxingdong he will once he finishes it.

**Comment by zuoxingdong** [comment]
@soumith Okay, thanks

**Comment by fmassa** [comment]
@zuoxingdong I'll try to work on it this weekend.

**Comment by zuoxingdong** [comment]
@fmassa  Great, looking forward to that! If I got it correctly, it shall be an API e.g. `a = gather_n()` to make it being able to backprop gradients with `a.data, a.grad` ? Or will it be 'Pythonic' e.g. directly with 
```
a = torch.ones([5, 4, 3])
slice = torch.LongTensor([0, 1, 1, 2, 0])
a_ = a[torch.range(5), :, slice]
a_.data
a_.grad
```

**Comment by fmassa** [comment]
@zuoxingdong it will be 'pythonic', accessing everything via `__getitem__`

**Comment by fmassa** [comment]
and for the time being, it will only support indexing, not assignment, so something like
```python
a[:5, torch.LongTensor([3, 2]), torch.LongTensor([1, 2])] = 1
```
won't be supported at first.

**Comment by zuoxingdong** [comment]
@fmassa  Sorry for a bit off this issue. I failed quite a while to use `gather()` for this purpose. Do you have some ideas how this going to be work, it is also on [Discuss forum](https://discuss.pytorch.org/t/4d-tensor-indexing-with-gather/1320). 

```
a = np.random.randn(5, 4, 3, 3)
a[range(5), :, [0, 1, 1, 2, 0], [1, 2, 0, 1, 0]]
```

**Comment by fmassa** [comment]
@zuoxingdong I replied in the thread.

**Comment by yuyu2172** [comment]
Recently, advanced indexing was added as a feature to Chainer (https://github.com/pfnet/chainer/pull/2203).
Hope that helps.

**Comment by fmassa** [comment]
@killeent here is the proof of concept implementation that I had for advanced indexing https://gist.github.com/fmassa/f8158d1dfd25a8047c2c668a44ff57f4

The idea was to convert the list of indices into a single linear index, `view` the tensor accordingly and then use `index_select` on the `viewed` tensor.

The current implementation in the gist supposes that only the indexing on the integer arrays are performed, and slicing / selection are not taken into account, but it should be easy to extend.

**Comment by killeent** [comment]
@fmassa - can you explain what subset of advanced indexing you are supporting here? I'm not sure what you mean by `convert the list of indices into a single linear index`

**Comment by fmassa** [comment]
@killeent This code supports advanced indexing as described in https://docs.scipy.org/doc/numpy/reference/arrays.indexing.html#integer-array-indexing 

Combining advanced and basic indexing is supported with the same behavior as numpy, the restriction at the moment is that I don't take into account partial slices (`a[1:3]`), for example
```python
x = np.random.random((10, 20, 30, 40, 50))
x[:, [1,2], :, [2,3], [1, 3]] # supported in the snippet, no partial slice
x[1:2, [1,2], 1:3, [2,3], [1, 3]] # not supported because of slices
```
In order to support partial slicing, what I'd do is to first perform the partial slicing over all dimensions, and then apply the advanced indexing. Something like
```python
# I want x[1:2, [1,2], 1:3, [2,3], [1, 3]]
# first perform basic indexing
temp = x[1:2, :, 1:3, :, :]
# now apply advanced indexing
x_final = temp[:, [1,2], :, [2,3], [1, 3]]
```

In the bottom of the gist there are some examples comparing the implementation with numpy one.

The idea of the linear index can be better explained with an example: for a tensor `x` of size `(5, 5)`, say we want to index it following `x[[0, 1], [1, 2]]`. This is equivalent to view `x` as a 1d tensor, and perform an `index_select` using as indices the linear indices corresponding to `[[0, 1], [1, 2]]`, which for a `5x5` tensor is `[1, 7]`.

Please let me know if I wasn't clear.

**Comment by killeent** [comment]
Cool. Thanks for the extended explanation @fmassa - can you extrapolate a little more on the "gaps in indices" part of the code: https://gist.github.com/fmassa/f8158d1dfd25a8047c2c668a44ff57f4#file-pytorch_advanced_indexing-py-L34

The explanation for combining advanced and basic indexing in the numpy documentation leaves a lot to be desired :)

**Comment by fmassa** [comment]
Sure, no problem. Indeed, it took me sometime to understand this indexing part in numpy.

I think it's easier to use an example here.
Say we have 
```python
x = np.random.random((10, 20, 30, 40, 50))
idx1 = [1, 8, 5]
idx2 = [4, 1, 6]
```
There are two possible cases: `idx1` and `idx2` appear in consecutive dimensions of the tensor. In this case, the position of the resulting tensor after indexing is well defined, and appears at the same position as the first index. So `x[:, idx1, idx2, :, :].shape == (10, 3, 40, 50)`.

Now, if the indices does not appear in consecutive dimensions (for example in `x[:, idx1, :, idx2, :]`), there is no unambiguous place to drop the indexing subspace, so the subspace is sent to the beginning of the array (yielding an array of shape `x[:, idx1, :, idx2, :].shape == (3, 10, 30, 50)`.

The code that you pointed out is doing just that. It checks if the indices appears in consecutive dimensions or not, and if not then permute the tensor so that the indexing dimensions comes first.

Let me know if you have other questions. We could also chat though slack if you prefer.

**Comment by killeent** [comment]
Discussed with @fmassa on Slack. I think the steps to complete this task are:

1. The simplest case of advanced indexing is having as input ndim(indexed tensor) input arrays. We can support this limited use case by verifying the input, and then porting the Python code above where we use `index_select` and `view` operations to generate the result.
2. Then, we can extend support to cases where we only perform advanced indexing on a subset of the tensor, with adjacent advanced indices, i.e. the case `x[:, idx1, idx2, :, :]` above
3. Extend the advanced indexing to support non-adjacent advanced indexing: `x[:, idx1, :, idx2, :]`
4. Implement 2-pass indexing to support the combination of basic indexing and advanced indexing: `x[1:2, [1,2], 1:3, [2,3], [1, 3]]`

I will do my best to tackle 1), and then will evaluate from there.

**Comment by xuancong84** [comment]
I think even Numpy's way advanced indexing is not very logical.
For the example,
```
x = np.random.random((10, 20, 30, 40, 50))
idx1 = [1, 8, 5]
idx2 = [4, 1, 6, 2]
```
**Logically speaking**, x[:, idx1, idx2, :, :] should have a shape of (10, 3, 4, 40, 50), and x[:, idx1, :, idx2, :] should have a shape of (10, 3, 30, 4, 50). The behaviour should be consistent regardless whether idx1 and idx2 are adjacent or not. For single entry dimension, x[:, 3, :, 3, :] should have a shape of (10, 30, 50) while x[:, [3], :, [3], :] should have a shape of (10, 1, 30, 1, 50), containing essentially the same data array. The idea is that for each dimension, if the index is an integer, then extract that 1-layer slice and perform a squeeze; if the index is a list, then extract that slice without performing squeeze regardless of whether the extracted slice has a thickness of 1 or not.

Then we consider the case if the number of indexed dimensions is fewer than the rank of the tensor. For example, x[:, idx1, :, ] should have a shape of (10, 3, 30, 40, 50) where you match dimension from the left; x[, :, idx2, :] should have a shape of (10, 20, 30, 4, 50) where you match dimension from the right. And x[:, idx1, :] should throw exception (or follow the match from left convention).

I am not sure what those Numpy guys are thinking about when they design the current indexing method.

**Comment by fmassa** [comment]
@xuancong84 the behavior of numpy makes sense to me, and what you would like to do is still possible with current implementation.
You should do the indexing twice, `x[:, idx1][:, :, idx2]`.

**Comment by killeent** [comment]
Yeah @xuancong84 the non-adjacent advanced indexing logic takes a while to wrap your head around, but it does make sense as to why the issue exists, and its largely to do with the fact that you can insert an arbitrary number of dimensions via your indexing.

**Comment by xuancong84** [comment]
Actually, I do fully understand the non-adjacent advanced indexing logic. But I don't think its design is good enough.
Consider the following example,
```
x = np.random.random((10, 20, 30, 40, 50))
idx1 = [1, 8, 5]
idx2 = [4, 1, 6]
```
In the current indexing logic design, the very fact that it matters whether the two indexing arrays (idx1 and idx2) are adjacent to each other or not, is because it could mean two things when multiple indexing arrays are provided:
1). **slicing**. x[:, idx1, idx2, :, :] could mean that along the 2nd dimension, we choose slices with indices idx1, and along the 3rd dimension, we choose slices with indices idx2.
2). **multi-point selection**. x[:, idx1, idx2, :, :] could mean that in the subspace along the 2nd and 3rd dimension, we choose points [(1,4), (8,1), (5,6)], which are in fact [(idx1[0], idx2[0]), (idx1[1], idx2[1]), (idx1[2], idx2[2])]. Therefore, idx1 and idx2 must have the same number of elements. (In fact, it will throw exception if idx1 and idx2 have different number of elements.)

Do take note that the difference between 1) and 2) is that:
**In 1)**, all the len(idx1)*len(idx2)=9 points are chosen. The subspace along the 2nd and 3rd dimension is a 20-by-30 grid plane, choosing [1, 8, 5] along the 2nd dimension (with max size 20), and choosing [4, 1, 6] along the 3rd dimension (with max size 30), this will end up with all the 3x3=9 points [(1,4), (1,1), (1,6), (8,4), (8,1), (8,6), (5,4), (5,1), (5,6)]. In this way, you cannot arbitrarily select any number of points in this subspace.
**In 2)**, you have the freedom to select any number of points in the 20-by-30 grid subspace. However, if you would like to choose N points along the 2nd dimension and M points along the 3rd dimension, then you need to list down all the N*M pairs of coordinates, i.e., len(idx1)=len(idx2)=N*M, which is very tedious.

The problem with 2) (the current indexing logic) is that it is neither self-consistent nor systematic. This is because even though idx1 and idx2 are not adjacent, you can still perform multi-point selection. And there is no reason why you must order the multi-point-selection dimension to the beginning just because idx1 and idx2 are adjacent. Instead, you could adopt either one of the following two conventions:
1. always bring to the front, e.g., 
```
x[:, idx1, idx2, :, :].shape=(3, 10, 40, 50)
x[:, idx1, :, idx2, :].shape=(3, 10, 30, 50)
```
2. stick to the 1st multi-point-selection coordinate dimension, e.g.,
```
x[:, idx1, idx2, :, :].shape=(10, 3, 40, 50)
x[:, idx1, :, idx2, :].shape=(10, 3, 30, 50)
x[:, :, idx1, idx2, :].shape=(10, 20, 3, 50)
x[:, :, idx1, :, idx2].shape=(10, 20, 3, 40)
```

A more elegant design would be the following, use list for **slicing** and use tuple for **multi-point selection**, or the other way round if you do not want to break compatibility with previous version. And adopt Convention 1, always bring the multi-point-selection dimension to the front, because the two can co-exist, and you don't want to mess up.

**Comment by CuriousCat-7** [comment]
Does anybody know how to do batch_index_select?

I want to slice cross different batches, and in every batch has a different index. What should I do?

Thank you

**Comment by gurkirt** [comment]
Hi Guys, I am also dealing with advanced indexing. I understand index select part and it works on single GPU I have the problem when it comes to multiple GPUs. Can you please have a look at https://discuss.pytorch.org/t/runtimeerror-index-select-with-dataparallel-arguments-are-located-on-different-gpus/10229
